File: vibecheck-core/tests/fixtures/lru_cache/human.js
// LRU cache for the session ctx layer
// ref: https://issues.internal/browse/FE-1092
// @tchen - let me know if we need thread safety here

/* eslint-disable no-underscore-dangle */

class LRUCache {
  constructor(cap) {
    this.cap = cap;
    this.sz = 0;
    this._map = new Map();
    // doubly-linked list w/ sentinel nodes
    this._head = { k: null, v: null, prv: null, nxt: null };
    this._tail = { k: null, v: null, prv: null, nxt: null };
    this._head.nxt = this._tail;
    this._tail.prv = this._head;
  }

  get(k) {
    const n = this._map.get(k);
    if (!n) return undefined;
    this._detach(n);
    this._pushFront(n);
    return n.v;
  }

  put(k, v) {
    if (this._map.has(k)) {
      const n = this._map.get(k);
      // const old = n.v;
      n.v = v;
      this._detach(n);
      this._pushFront(n);
      return;
    }

    // TODO: maybe batch evictions when sz >> cap? perf tbd
    if (this.sz >= this.cap) {
      this._evict();
    }

    const n = { k, v, prv: null, nxt: null };
    this._map.set(k, n);
    this._pushFront(n);
    this.sz++;
  }

  _pushFront(n) {
    const x = this._head.nxt;
    this._head.nxt = n;
    n.prv = this._head;
    n.nxt = x;
    x.prv = n;
  }

  _detach(n) {
    const p = n.prv;
    const x = n.nxt;
    p.nxt = x;
    x.prv = p;
  }

  // FIXME: should we emit an event on eviction? see #789
  _evict() {
    if (this.sz === 0) return;
    const t = this._tail.prv;
    this._detach(t);
    this._map.delete(t.k);
    // console.log(`evicted: ${t.k}`);
    this.sz--;
  }

  has(k) {
    return this._map.has(k);
  }

  keys() {
    const res = [];
    let c = this._head.nxt;
    while (c !== this._tail) {
      res.push(c.k);
      c = c.nxt;
    }
    return res;
  }

  get size() {
    return this.sz;
  }

  // eslint-disable-next-line class-methods-use-this
  clear() {
    this._map.clear();
    this._head.nxt = this._tail;
    this._tail.prv = this._head;
    this.sz = 0;
  }
}

module.exports = { LRUCache };

File: vibecheck-core/tests/fixtures/lru_cache/human.py
"""Simple LRU cache implementation.
See ticket PROJ-287 for context on why we rolled our own.
@jpark reviewed 2024-11-03
"""


class _Node:
    __slots__ = ("k", "v", "prv", "nxt")

    def __init__(self, k, v):  # type: ignore
        self.k = k
        self.v = v
        self.prv = None  # type: ignore
        self.nxt = None  # type: ignore


class LRUCache:
    def __init__(self, cap: int):
        # TODO: add ttl support eventually (see #221)
        self.cap = cap
        self.sz = 0
        self._map: dict = {}
        # sentinel nodes to avoid null checks
        self._head = _Node("", None)
        self._tail = _Node("", None)
        self._head.nxt = self._tail
        self._tail.prv = self._head

    def get(self, k):
        n = self._map.get(k)
        if n is None:
            return None
        self._detach(n)
        self._push_front(n)
        return n.v

    def put(self, k, v):  # noqa: C901
        if k in self._map:
            n = self._map[k]
            # old = n.v
            n.v = v
            self._detach(n)
            self._push_front(n)
            return

        if self.sz >= self.cap:
            self._evict()

        n = _Node(k, v)
        self._map[k] = n
        self._push_front(n)
        self.sz += 1

    def _push_front(self, n):
        x = self._head.nxt
        self._head.nxt = n
        n.prv = self._head
        n.nxt = x
        x.prv = n  # type: ignore

    def _detach(self, n):
        p = n.prv
        x = n.nxt
        p.nxt = x
        x.prv = p

    # FIXME: doesn't handle edge case where cap is changed at runtime
    def _evict(self):
        if self.sz == 0:
            return
        t = self._tail.prv
        # print(f"evicting key={t.k}")
        self._detach(t)
        del self._map[t.k]  # type: ignore
        self.sz -= 1

    def keys(self):
        """Iterate keys in MRU order."""
        cur = self._head.nxt
        while cur is not self._tail:
            yield cur.k
            cur = cur.nxt

    def __len__(self):
        return self.sz

    def __contains__(self, k):
        return k in self._map

    def __repr__(self):
        items = [(n.k, n.v) for n in self._map.values()]
        return f"LRUCache(cap={self.cap}, items={items})"
